File: src/main/java/no/fintlabs/dispatch/WebClientRequestService.java
package no.fintlabs.dispatch;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.dispatch.model.InstanceReceiptDispatchEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Service
@Slf4j
public class WebClientRequestService {
    private final WebClient webClient;

    public WebClientRequestService(WebClient webClient) {
        this.webClient = webClient;
    }

    public Mono<InstanceReceiptDispatchEntity> dispatchInstance(
            InstanceReceiptDispatchEntity instanceReceiptDispatchEntity
    ) {
        ObjectMapper objectMapper = new ObjectMapper();
        Object instanceToDispatch;
        try {
            instanceToDispatch = objectMapper.readValue(
                    instanceReceiptDispatchEntity.getInstanceReceipt(), instanceReceiptDispatchEntity.getClassType()
            );
        } catch (JsonProcessingException e) {
            return Mono.error(e);
        }

        return webClient.patch()
                .uri(instanceReceiptDispatchEntity.getUri())
                .body(Mono.just(instanceToDispatch), instanceReceiptDispatchEntity.getClassType())
                .retrieve()
                .bodyToMono(String.class)
                .doOnSubscribe(sub -> log.info("Dispatching to URI: {}", instanceReceiptDispatchEntity.getUri()))
                .doOnSuccess(response -> log.info("success {}", response))
                .doOnError(error -> log.error("Error msg from webclient: {}", error.getMessage()))
                .thenReturn(instanceReceiptDispatchEntity);
    }


}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/dispatch/WebClientRequestService.kt
package no.novari.flyt.egrunnerverv.gateway.dispatch

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import no.novari.flyt.egrunnerverv.gateway.dispatch.model.InstanceReceiptDispatchEntity
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClient
import org.springframework.web.client.RestClientResponseException

@Service
class WebClientRequestService(
    @param:Qualifier("dispatchRestClient") private val restClient: RestClient,
) {
    private val objectMapper = ObjectMapper()

    fun dispatchInstance(instanceReceiptDispatchEntity: InstanceReceiptDispatchEntity): InstanceReceiptDispatchEntity {
        val classType =
            instanceReceiptDispatchEntity.classType
                ?: throw IllegalStateException(
                    "Missing classType for ${instanceReceiptDispatchEntity.sourceApplicationInstanceId}",
                )

        val instanceToDispatch =
            try {
                objectMapper.readValue(
                    instanceReceiptDispatchEntity.instanceReceipt,
                    classType,
                )
            } catch (e: JsonProcessingException) {
                throw e
            }

        try {
            restClient
                .patch()
                .uri(instanceReceiptDispatchEntity.uri)
                .body(instanceToDispatch)
                .retrieve()
                .toBodilessEntity()
        } catch (ex: RestClientResponseException) {
            throw ex
        }

        return instanceReceiptDispatchEntity
    }
}
